zebra-network/src/peer_set/set/tests.rs

```diff
@@ -0,0 +1,382 @@
+    ) -> Arc<std::sync::Mutex<AddressBook>> {
+        let address_book = maybe_address_book.unwrap_or_else(Self::fallback_address_book);
+
+        self.address_book = Some(address_book.clone());
+
+        address_book
+    }
+
+    /// Create an empty [`AddressBook`] instance using a dummy local listener address.
+    fn fallback_address_book() -> Arc<std::sync::Mutex<AddressBook>> {
+        let local_listener = "127.0.0.1:1000"
+            .parse()
+            .expect("Invalid local listener address");
+        let address_book = AddressBook::new(local_listener, Span::none());
+
+        Arc::new(std::sync::Mutex::new(address_book))
+    }
+}
+
+/// A pair of block height values, where one is before and the other is at or after an arbitrary
+/// network upgrade's activation height.
+#[derive(Clone, Copy, Debug)]
+pub struct BlockHeightPairAcrossNetworkUpgrades {
+    /// The network for which the block height values represent heights before and after an
+    /// upgrade.
+    pub network: Network,
+
+    /// The block height before the network upgrade activation.
+    pub before_upgrade: block::Height,
+
+    /// The block height at or after the network upgrade activation.
+    pub after_upgrade: block::Height,
+}
+
+impl Arbitrary for BlockHeightPairAcrossNetworkUpgrades {
+    type Parameters = ();
+
+    fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
+        any::<(Network, NetworkUpgrade)>()
+            // Filter out genesis upgrade because there is no block height before genesis.
+            .prop_filter("no block height before genesis", |(_, upgrade)| {
+                !matches!(upgrade, NetworkUpgrade::Genesis)
+            })
+            // Filter out network upgrades without activation heights.
+            .prop_filter_map(
+                "missing activation height for network upgrade",
+                |(network, upgrade)| {
+                    upgrade
+                        .activation_height(network)
+                        .map(|height| (network, height))
+                },
+            )
+            // Obtain random heights before and after (or at) the network upgrade activation.
+            .prop_flat_map(|(network, activation_height)| {
+                let before_upgrade_strategy = 0..activation_height.0;
+                let after_upgrade_strategy = activation_height.0..;
+
+                (
+                    Just(network),
+                    before_upgrade_strategy,
+                    after_upgrade_strategy,
+                )
+            })
+            // Collect the arbitrary values to build the final type.
+            .prop_map(|(network, before_upgrade_height, after_upgrade_height)| {
+                BlockHeightPairAcrossNetworkUpgrades {
+                    network,
+                    before_upgrade: dbg!(block::Height(before_upgrade_height)),
+                    after_upgrade: dbg!(block::Height(after_upgrade_height)),
```

Comment:
Nit: Remove `dbg!` macros

```suggestion
                    before_upgrade: block::Height(before_upgrade_height),
                    after_upgrade: block::Height(after_upgrade_height),
```